### PR TITLE
:tada: Migrate repo to Kubebuilder v2

### PR DESCRIPTION
**Pull request by vincepri**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -20,3 +21,24 @@ workspace_binary(
...
+        "//controllers:go_default_library",
+        "//vendor/k8s.io/apimachinery/pkg/runtime:go_default_library",
+        "//vendor/k8s.io/client-go/kubernetes/scheme:go_default_library",
+        "//vendor/k8s.io/klog/klogr:go_default_library",
+        "//vendor/sigs.k8s.io/controller-runtime:go_default_library",
+    ],
+)
+
+go_binary(
+    name = "cluster-api",
```
Comment (detiber) [inline BUILD.bazel]: might be better to call this "cluster-api-manager" instead.

File: Makefile [hunk 2]
```diff
@@ -107,10 +107,10 @@ generate-deepcopy: ## Runs controller-gen to generate deepcopy files
 .PHONY: generate-manifests
 generate-manifests: ## Generate manifests e.g. CRD, RBAC etc.
 	go run vendor/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go \
-		paths=./pkg/... \
+		paths=./api/... \
```
Comment (detiber) [inline Makefile]: shouldn't this also include the controllers directory as well?

File: api/v1alpha2/cluster_types.go [hunk 3]
```diff
@@ -22,30 +22,13 @@ import (
 	capierrors "sigs.k8s.io/cluster-api/pkg/errors"
 )
 
+// +genclient
```
Comment (detiber) [inline api/v1alpha2/cluster_types.go]: Do we need genclient? I thought we were moving towards just the controller-runtime client.

File: api/v1alpha2/machinedeployment_types.go [hunk 4]
```diff
@@ -21,6 +21,8 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
+// +genclient
```
Comment (detiber) [inline api/v1alpha2/machinedeployment_types.go]: Any reason we still need genclient here?

File: api/v1alpha2/machineset_types.go [hunk 5]
```diff
@@ -27,24 +27,6 @@ import (
 )
 
 // +genclient
```
Comment (detiber) [inline api/v1alpha2/machineset_types.go]: Is genclient still needed here?

File: Makefile (hunk 2 above)
Comment (vincepri) [reply Makefile]: Added!

**Comment by vincepri** [comment]
@detiber fixed all comments 🙂

File: config/crd/bases/cluster.x-k8s.io_machines.yaml [hunk 6]
```diff
@@ -254,7 +254,8 @@ spec:
           - infrastructureRef
           type: object
         status:
-          description: MachineStatus defines the observed state of Machine
+          description: / [MachineStatus] MachineStatus defines the observed state
```
Comment (detiber) [inline config/crd/bases/cluster.x-k8s.io_machines.yaml]: This generation looks a bit off.

File: config/crd/bases/cluster.x-k8s.io_machinesets.yaml [hunk 7]
```diff
@@ -21,8 +21,7 @@ spec:
     status: {}
   validation:
     openAPIV3Schema:
-      description: / [MachineSet] MachineSet ensures that a specified number of machines
-        replicas are running at any given time.
+      description: / [MachineSet] MachineSet is the Schema for the machinesets API
```
Comment (detiber) [inline config/crd/bases/cluster.x-k8s.io_machinesets.yaml]: Same here with the generation looking off

File: controllers/cluster_controller.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/status,verbs=get;update;patch
+
+func (r *ClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("cluster", req.NamespacedName)
+
+	// your logic here
```
Comment (detiber) [inline controllers/cluster_controller.go]: Still a todo to migrate the controller logic here?

File: controllers/machine_controller.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines/status,verbs=get;update;patch
+
+func (r *MachineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("machine", req.NamespacedName)
+
+	// your logic here
```
Comment (detiber) [inline controllers/machine_controller.go]: Still a todo to migrate the controller logic here?

File: controllers/machinedeployment_controller.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments/status,verbs=get;update;patch
+
+func (r *MachineDeploymentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("machinedeployment", req.NamespacedName)
+
+	// your logic here
```
Comment (detiber) [inline controllers/machinedeployment_controller.go]: Still a todo to migrate the controller logic here?

File: controllers/machineset_controller.go [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinesets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinesets/status,verbs=get;update;patch
+
+func (r *MachineSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("machineset", req.NamespacedName)
+
+	// your logic here
```
Comment (detiber) [inline controllers/machineset_controller.go]: Still a todo to migrate the controller logic here?

File: pkg/provider/example/container/Dockerfile [hunk 12]
```diff
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 # Build the manager binary
-FROM golang:1.12.5 as builder
+FROM golang:1.12.6 as builder
```
Comment (detiber) [inline pkg/provider/example/container/Dockerfile]:
```suggestion
FROM golang:1.12.7 as builder
```

File: controllers/machinedeployment_controller.go (hunk 10 above)
Comment (vincepri) [reply controllers/machinedeployment_controller.go]: Yeah, it's gonna be a different PR

File: config/crd/bases/cluster.x-k8s.io_machines.yaml (hunk 6 above)
Comment (vincepri) [reply config/crd/bases/cluster.x-k8s.io_machines.yaml]: This has been shown before as well, we were missing marks on the resources, I added them back given that we'll probably reference these from the gitbook

**Comment by detiber** [comment]
lgtm, @chuckha mind giving this a second set of eyes?

File: api/v1alpha2/cluster_types.go [hunk 13]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Kubernetes Authors.
+Copyright The Kubernetes Authors.
```
Comment (chuckha) [inline api/v1alpha2/cluster_types.go]: This is going to need the year in the copyright even though it's a generated file it's not meant to be re-generated. This is true for all the generated files not destined to be regularly-regenerated below as well. And for the actual generated files (deep copy) you will need to either: 1) extend the regex in the python validation script or 2) add the words `DO NOT EDIT` to the file

Comment (vincepri) [reply api/v1alpha2/cluster_types.go]: Fixed the copyright in these files and the controllers/* ones. I opened https://github.com/kubernetes-sigs/controller-tools/pull/306 to address the deepcopy issue

Comment (chuckha) [reply api/v1alpha2/cluster_types.go]: much appreciated.

**Comment by vincepri** [comment]
@detiber @chuckha Is this good to go? I'll start the work to migrate the controllers after this merges

**Comment by chuckha** [comment]
@vincepri squash your commits and lgtm

or if you don't want to squash i'll lgtm

**Comment by vincepri** [comment]
@chuckha We're not squashing commits anymore, at this within this repository, per https://github.com/kubernetes-sigs/cluster-api/issues/1191, given the large footprint of the PR, I think the commits should be ok?

**Comment by chuckha** [comment]
The idea behind that change was that people are responsible for squashing their own commits instead of the merge bot. It was not that we would keep all commits that the dev made while developing because they are usually not particularly meaningful. The general practice is 1 commit for changes 1 for generated file updates, however, if you're happy with these commits no way I'd block on it

**Comment by chuckha** [comment]
/lgtm

Thanks!

**Pull request merged**